xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java

```diff
@@ -0,0 +1,196 @@
+
+        final AtomicBoolean completed = new AtomicBoolean();
+        final CompletableFuture<List<Endpoint>> whenReady = kubernetesEndpointGroup.whenReady();
+        final ServiceRequestContext ctx = ServiceRequestContext.current();
+
+        // Use a schedule to time out the watcher creation until we implement a custom implementation.
+        final ScheduledFuture<?> scheduledFuture = ctx.eventLoop().schedule(() -> {
+            if (!completed.compareAndSet(false, true)) {
+                return;
+            }
+            kubernetesEndpointGroup.closeAsync();
+            responseObserver.onError(
+                    Status.INTERNAL.withDescription(
+                            "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
+                            watcher.getName()).asRuntimeException());
+        }, 5, TimeUnit.SECONDS);
+
+        whenReady.handle((endpoints, cause) -> {
+            if (!completed.compareAndSet(false, true)) {
+                return null;
+            }
+            scheduledFuture.cancel(false);
+            kubernetesEndpointGroup.closeAsync();
+            if (cause != null) {
+                // Specific types.
+                responseObserver.onError(Status.INTERNAL.withCause(cause).asRuntimeException());
+                return null;
+            }
+            logger.debug("Successfully retrieved k8s endpoints: {}", endpoints);
+            onSuccess.run();
+            return null;
+        });
+    }
+
+    /**
+     * Creates a {@link KubernetesEndpointGroup} from the specified {@link Watcher}.
+     * This method must be executed in a blocking thread because
+     * {@link KubernetesEndpointGroupBuilder#build()} blocks the execution thread.
+     */
+    public static KubernetesEndpointGroup createKubernetesEndpointGroup(Watcher watcher) {
+        final KubernetesConfig kubernetesConfig = watcher.getKubernetesConfig();
+        final String serviceName = watcher.getServiceName();
+
+        final KubernetesEndpointGroupBuilder kubernetesEndpointGroupBuilder =
+                KubernetesEndpointGroup.builder(toConfig(kubernetesConfig)).serviceName(serviceName);
+        if (!isNullOrEmpty(kubernetesConfig.getNamespace())) {
+            kubernetesEndpointGroupBuilder.namespace(kubernetesConfig.getNamespace());
+        }
+        if (!isNullOrEmpty(watcher.getPortName())) {
+            kubernetesEndpointGroupBuilder.portName(watcher.getPortName());
+        }
+
+        return kubernetesEndpointGroupBuilder.build();
+    }
+
+    private static Config toConfig(KubernetesConfig kubernetesConfig) {
+        final ConfigBuilder configBuilder = new ConfigBuilder()
+                .withMasterUrl(kubernetesConfig.getControlPlaneUrl())
+                .withTrustCerts(kubernetesConfig.getTrustCerts());
+
+        if (!isNullOrEmpty(kubernetesConfig.getOauthToken())) {
+            configBuilder.withOauthToken(kubernetesConfig.getOauthToken());
```

Comment:
Question) Am I understanding correctly that as of now users can access this value via APIs? I'm wondering if we're going need to prioritize access control for xDS related APIs.

Reply:
Users cannot see this value. Only creating, updating and deleting is possible at the moment.